File: api/schemas/scoreboard.py
"""Schemas served by the Scoreboard summary."""

from datetime import date
from pydantic import BaseModel, Field
from api.schemas.common import BootstrapSectionStatus, SourceDescriptor

# ---- /scoreboard/summary weekly ------------------------------------------
#
# The full weekly backtest series plus independently hours-weighted pools. All
# spans all time; pre-RTC+B is backtest-only and post-RTC+B includes live days.


class WeeklyPoint(BaseModel):
    """One ``scoreboard_weekly`` row served for the chart — one (week, source).

    Screening metrics (``rank_spearman``/``sign_agree``/``topdecile_hit``) lead
    the page. All nullable — a NULL cell (e.g. the flat ``null`` source's
    declined top-decile) rides through as ``None``.

    """

    week: date
    source_id: str
    series_id: str
    rank_spearman: float | None = None
    sign_agree: float | None = None
    topdecile_hit: float | None = None
    sf_coverage: float | None = None
    model_coverage: float | None = None
    n_hours: int | None = None
    n_nodes: int | None = None


class SourcePooled(BaseModel):
    """One source's hours-weighted metrics over a split, or ``None`` if unscored."""

    source_id: str
    series_id: str
    rank_spearman: float | None = None
    sign_agree: float | None = None
    topdecile_hit: float | None = None


class WeeklySplit(BaseModel):
    """A pooled slice — ``all`` / ``pre_rtc_b`` / ``post_rtc_b`` — so the RTC+B
    structural break is visible on every pooled stat.

    ``beats_persistence`` is the existence test (model > persistence on all
    three screening currencies).

    ``sources`` always includes model + persistence + climatology + oracle

    """

    label: str  # all | pre_rtc_b | post_rtc_b
    n_weeks: int
    n_days: int
    sources: list[SourcePooled]
    beats_persistence: bool | None = None


class ScoreboardWeekly(BaseModel):
    """The weekly series + pooled summary for one board.

    ``primary_source`` echoes the requested ``source`` (the series the page
    foregrounds) ``rtc_b_cutover`` is the split date the
    ``pre_``/``post_rtc_b`` summaries divide on.

    """

    run_id: str
    primary_source_id: str
    rtc_b_cutover: date
    points: list[WeeklyPoint]
    splits: list[WeeklySplit]
    sources: list[SourceDescriptor] = Field(default_factory=list)


# ---- /scoreboard/summary daily section -----------------------------------
#
# The LIVE scoreboard. per-delivery-day grades of the SERVED forecast, scored
# against realized once DAM publishes, the live counterpart to the weekly
# backtest board.
#

class DailyPoint(BaseModel):
    """One ``scoreboard_daily`` row — one (delivery_date, source) live grade.

    """

    delivery_date: date
    source_id: str
    series_id: str
    horizon: int
    rank_spearman: float | None = None
    sign_agree: float | None = None
    topdecile_hit: float | None = None
    sf_coverage: float | None = None
    model_coverage: float | None = None
    n_hours: int | None = None
    n_nodes: int | None = None


class ScoreboardDaily(BaseModel):
    """The newest final live grade for one run.

    ``run_id`` is the model version being graded live (the same version the
    forecast served).

    ``primary_source`` identifies the model series the page foregrounds; every
    source rides along in ``points`` regardless, so the client can never render
    a lone model figure.

    ``selected_delivery_date`` and ``horizon`` make the selected final forecast
    explicit.

    """

    run_id: str
    primary_source_id: str
    horizon: int
    selected_delivery_date: date
    points: list[DailyPoint]
    sources: list[SourceDescriptor] = Field(default_factory=list)


# ---- /scoreboard/summary history -----------------------------------------


class ScoreHistoryPoint(BaseModel):
    """One chart point from either separately returned Scoreboard cadence."""

    source_id: str
    series_id: str
    week: date | None = None
    delivery_date: date | None = None
    rank_spearman: float | None = None
    sign_agree: float | None = None
    topdecile_hit: float | None = None
    sf_coverage: float | None = None
    model_coverage: float | None = None
    n_hours: int | None = None
    n_nodes: int | None = None


class ScoreboardHistory(BaseModel):
    """Distinct weekly walk-forward and final served-day chart histories."""

    primary_source_id: str
    weekly_run_id: str
    daily_run_id: str | None = None
    weekly_points: list[ScoreHistoryPoint]
    served_daily_points: list[ScoreHistoryPoint] = Field(default_factory=list)
    sources: list[SourceDescriptor] = Field(default_factory=list)


class ScoreboardSummaryResponse(BaseModel):
    """One bundled payload for the Scoreboard page summary.

    ``weekly`` reads ``scoreboard_weekly``; ``daily`` is the latest final served
    grade and the final served tail of ``history`` resolves independently from
    ``scoreboard_daily``.

    A field is ``null`` exactly when its source section would 503.

    """

    weekly: ScoreboardWeekly | None
    daily: ScoreboardDaily | None
    history: ScoreboardHistory | None
    availability: dict[str, BootstrapSectionStatus]

File: web/src/api/types/scoreboard.ts
import type { BootstrapSectionStatus, SourceDescriptor } from "./common";

// =============================================================================
// /scoreboard/summary weekly — the full weekly backtest series + pooled pre/post-RTC+B
// summary (the scoreboard page's data). Mirrors api/models.py WeeklyPoint /
// SourcePooled / WeeklySplit / ScoreboardWeekly.
// =============================================================================

// One (week, source) row for the chart. All nullable — a declined screening
// metric rides through as null.
export interface WeeklyPoint {
  week: string;
  source_id: string;
  series_id: string;
  rank_spearman: number | null;
  sign_agree: number | null;
  topdecile_hit: number | null;
  coverage80: number | null;
  band_width: number | null;
  pinball: number | null;
  sf_coverage: number | null;
  model_coverage: number | null;
  n_hours: number | null;
  n_nodes: number | null;
}

// One source's independently hours-weighted metrics over a split.
export interface SourcePooled {
  source_id: string;
  series_id: string;
  rank_spearman: number | null;
  sign_agree: number | null;
  topdecile_hit: number | null;
}

// A pooled slice (all / pre_rtc_b / post_rtc_b) with an existence test.
export interface WeeklySplit {
  label: string; // all | pre_rtc_b | post_rtc_b
  n_weeks: number;
  n_days: number;
  sources: SourcePooled[];
  beats_persistence: boolean | null;
}

export interface ScoreboardWeekly {
  run_id: string;
  primary_source_id: string;
  rtc_b_cutover: string;
  points: WeeklyPoint[];
  splits: WeeklySplit[];
  sources: SourceDescriptor[];
}

// =============================================================================
// /scoreboard/summary daily section — the newest final LIVE grade.
// Mirrors api/models.py DailyPoint / ScoreboardDaily. The live
// counterpart to the weekly backtest board: per-day grades of the SERVED
// forecast, same currency columns as WeeklyPoint so a live number and a backtest
// number are directly comparable. Every response carries all sources (model +
// persistence + climatology + oracle + the `null` flat tripwire.)
// =============================================================================

// One (delivery_date, source) live grade. Band columns (coverage80 / band_width
// / pinball) are populated on the model source only; model_coverage is NULL for
// now (deferred snapshot).
export interface DailyPoint {
  delivery_date: string;
  source_id: string;
  series_id: string;
  // Which forecast track this row grades: 1 = final (fires D−1), 2 = preview
  // (fires D−2). One board carries one horizon; it rides on every point so a
  // reader never has to guess which track a number belongs to.
  horizon: number;
  rank_spearman: number | null;
  sign_agree: number | null;
  topdecile_hit: number | null;
  coverage80: number | null;
  band_width: number | null;
  pinball: number | null;
  sf_coverage: number | null;
  model_coverage: number | null;
  n_hours: number | null;
  n_nodes: number | null;
}

export interface ScoreboardDaily {
  run_id: string;
  primary_source_id: string;
  // The track `points` grades; this summary section is always the final track.
  horizon: number;
  selected_delivery_date: string;
  points: DailyPoint[];
  sources: SourceDescriptor[];
}

// The `/scoreboard/summary` chart sequence. Weekly backtest and daily served
// grades retain distinct date fields and provenance; the final track is the
// only live cadence included.
export type ScoreCadence = "backtest_weekly" | "served_daily";

export interface ScoreHistoryPoint {
  source_id: string;
  series_id: string;
  cadence: ScoreCadence;
  week: string | null;
  delivery_date: string | null;
  rank_spearman: number | null;
  sign_agree: number | null;
  topdecile_hit: number | null;
  sf_coverage: number | null;
  model_coverage: number | null;
  n_hours: number | null;
  n_nodes: number | null;
}

export interface ScoreboardHistory {
  primary_source_id: string;
  weekly_run_id: string;
  daily_run_id: string | null;
  boundary_date: string | null;
  points: ScoreHistoryPoint[];
  sources: SourceDescriptor[];
}

// /scoreboard/summary — one bundled payload for the Scoreboard page's load-time
// requests. `availability` makes every nullable section's soft-fail state
// explicit.
export interface ScoreboardSummary {
  weekly: ScoreboardWeekly | null;
  daily: ScoreboardDaily | null;
  history: ScoreboardHistory | null;
  availability: Record<string, BootstrapSectionStatus>;
}
